### PR TITLE
Add SERIAL_JTAG/CDC logger option for ESP-IDF platform for ESP32-S2/S3/C3

### LINKED ISSUE #1598: Add USB CDC support to Logger component
ESP32S2 and ESP32C3 has integrated USB support and modules usually do not include an external USB -> serial bridge. 
But EspHome does not support USB CDC, leaving no easy path forward. 

**Describe the problem you have/What new integration you would like**

Add support for USB CDC in the logger component, to make logging even simpler.

**Please describe your use case for this integration and alternatives you've tried:**

We can use a hardware UART, but it requires an extra USB to serial bridge and consumes one of the two UARTs in ESP32S2/C3. Some modules (like LOLIN S2 mini) not even expose the UART0 pins, leaving no hardware UARTs available.

**Additional context**

### PR DESCRIPTION
**Pull request by unaiur**
# What does this implement/fix? 

ESP-IDF configures a stdout device that is used for internal logs.
Esphome can write logs to stdout also to stdout and consolidate output.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes esphome/feature-requests#1598

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** 

PENDING

esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

Use USB CDC device in ESP32S2 chip:

```yaml
# Example config.yaml
esphome:
  name: test-s2

external_components:
  - source:
      type: local
      path: esphome/esphome/components
    components: [ logger ]

esp32:
  board: featheresp32-s2
  framework:
    type: esp-idf
    version: recommended
    sdkconfig_options:
      CONFIG_ESP_CONSOLE_USB_CDC: y

# Enable logging
logger:
  level: VERY_VERBOSE
  hardware_uart: stdout

# Enable Home Assistant API
api:
  password: ""

ota:
  password: !secret ota_password

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "test-s2"
    password: !secret ap_password

sensor:
  - platform: uptime
    name: Uptime Sensor
```

Using UART0 in ESP32 chip:
```
esphome:
  name: test-esp32

external_components:
  - source:
      type: local
      path: esphome/esphome/components
    components: [ logger ]

esp32:
  board: esp32doit-devkit-v1
  framework:
    type: esp-idf
    version: recommended


# Enable logging
logger:
  level: VERY_VERBOSE
  hardware_uart: STDOUT

# Enable Home Assistant API
api:
  password: ""

ota:
  password: !secret ota_password

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "test-esp32"
    password: !secret ap_password

sensor:
  - platform: uptime
    name: Uptime Sensor
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

PENDING

### CODE REVIEW

**Comment by unaiur** [comment]
This is a minimal approach of adding stdout as a hardware device. Main advantage is that, when using stdout, we get all the messages logged to the same device. Otherwise, early boot messages can go to another device.


But maybe it is better to replace completely the ESP-IDF implementation to always use stdout, and use the configuration flags to setup the right SDK config options. This will simplify logger implementation in ESP-IDF, but will also make more complex to provide alternative logging devices (they will have to redirect stdout).

I will prepare an alternative proposal using that approach.

**Comment by OttoWinter** [comment]
Tbh I like the cdc option more.

What's stdout supposed to mean for a user? As a user I just think I want the logging go to this uart interface, or the usb port.

This option value would be more for developers. So maybe we could offer it _additionally_ to a CDC option (but only if there are apis to intercept stdout for developers/stdout offers more functionality)

**Comment by unaiur** [comment]
> Tbh I like the cdc option more.
> 
> What's stdout supposed to mean for a user? As a user I just think I want the logging go to this uart interface, or the usb port.
> 
> This option value would be more for developers. So maybe we could offer it _additionally_ to a CDC option (but only if there are apis to intercept stdout for developers/stdout offers more functionality)

Yes, I also think that stdout is an implementation detail. We should configure ESP-IDF to setup stdout/stderr correctly and then log to stdout/stderr always.

Also, S2 chip can use two USB stacks: one integrated in the ROM that only supports CDC and DFU protocols and tinyUSB stack that is more flexible and is stored in the flash.

I would like to support both stacks, not only the ROM stack. But current SDK config option only works with the ROM USB stack.

**Comment by jbergler** [comment]
I suspect I'm doing something wrong, but I didn't have any luck with this on a C3.

I tried both `CONFIG_ESP_CONSOLE_USB_CDC` and `CONFIG_ESP_CONSOLE_USB_SERIAL_JTAG` with no change in behavior with the esp-idf 4.3. I tried to get both latest dev, and 4.4 from the WIP PR on the platformio framework but couldn't succeed in getting either of those to compile.

Happy to try and help test things if you like @unaiur, otherwise I'll keep tinkering to see if I can figure it out.

**Comment by jbergler** [comment]
> I suspect I'm doing something wrong, but I didn't have any luck with this on a C3.

As suspected user error: `version: latest` is 4.3.0 not 4.3.2 where the support was added.
[Added support for console output via USB_SERIAL_JTAG port on ESP32-C3](https://github.com/espressif/esp-idf/releases/tag/v4.3.2)

Targeting `version: 4.3.2` explicitly makes things work

**Comment by ashald** [comment]
FWIW, I tried it with Adafruit ESP32-S2 feather and it worked beautifully.
The only issue I noticed is that when I quite the `logs` command - the MCU restarts. o_O

**Comment by LeDominik** [comment]
🥳 success here as well on an [Unexpected Maker FeatherS2 Neo](https://unexpectedmaker.com/feathers2-neo), works like a charm. It also reboots when I disconnect but hey...I'm not going to complain, it works. Used the pio board `featheresp32-s2` and of course `CONFIG_ESP_CONSOLE_USB_CDC: y` in the IDF `sdkconfig_options`. Really looking forward to trying it on the ESP32S3 (still somewhere in the mail) which then also has JTAG over USB...

What I'm wondering about is how to have a _streamlined_ configuration that everyone understands _without_ grokking ESP IDF, stdout and so on. 2 thoughts:

- simplest would be to have some kind of "use in-chip USB" (for chips that can do it, so right now S2/C3/S3) thing within the `esp32` config that turns on all the knobs to use IDF plus the ROM-based USB. On S3 it would also activate JTAG over USB I guess. This could become **pretty standard**, if you browse around there's a clear trend to skip the extra chips 😄
- for slightly more advanced users I also think that when you activate IDF you really want STDOUT logging irrespective of where it then leads to

Tactically (if I had a say) the STDOUT logging here should be merged in ASAP with a minor release. Any "nice configuration" (for e.g. a major release) builds on this foundation.

**Comment by ashald** [comment]
@jbergler could you share your config for C3 please? This didn't work for me for c3
```
esphome:
  name: c3
  platformio_options:
    board_build.flash_mode: dio

esp32:
  board: esp32-c3-devkitm-1
  framework:
    type: esp-idf
    version: 4.3.2
    sdkconfig_options:
      CONFIG_ESP_CONSOLE_USB_CDC: y

external_components:
  - source: github://unaiur/esphome@dev-stdout-logger
    components:
      - logger

logger:
  hardware_uart: stdout
```
while a similar config worked for S2.

**Comment by jbergler** [comment]
@ashald try this
```

esphome:
  platformio_options:
    board_build.flash_mode: dio

esp32:
  board: esp32-c3-devkitm-1
  framework:
    type: esp-idf
    version: 4.3.2
    sdkconfig_options:
      CONFIG_ESP_CONSOLE_USB_SERIAL_JTAG: y

external_components:
  - source: github://unaiur/esphome@dev-stdout-logger
    components:
      - logger
 
logger:
  hardware_uart: STDOUT
```

**Comment by ashald** [comment]
@jbergler yay, it works! Thanks a lot!

**Comment by jhamhader** [comment]
Confirming `hardware_uart: STDOUT` with `CONFIG_ESP_CONSOLE_USB_CDC` works (edit: on S2).\
Thanks @unaiur!

While `hardware_uart: STDOUT` is technically correct, it is quite confusing.\
I'm wondering if `hardware_uart: CDC` and `hardware_uart: SERIAL_JTAG` (implying stdout and `CONFIG_ESP_CONSOLE_USB_CDC` or `CONFIG_ESP_CONSOLE_USB_SERIAL_JTAG`) could be easier for users and decouple the implementation (stdout in esphome + flag to the framework) from the intent.\
Verification can be placed to ensure this works only on relevant platforms.

**Comment by jesserockz** [comment]
> I'm wondering if hardware_uart: CDC and hardware_uart: SERIAL_JTAG (implying stdout and CONFIG_ESP_CONSOLE_USB_CDC or CONFIG_ESP_CONSOLE_USB_SERIAL_JTAG) could be easier for users

This I like. We could automatically add the correct sdkconfig flag based on this too.

**Comment by LeDominik** [comment]
@jhamhader Love setting it as CDC, just like @jesserockz  I find that super clear

**Comment by adonno** [comment]
not sure if it's related but i'm getting the following. 

[00:52:56][D][esp-idf:000]: E (34658) uart: uart_get_buffered_data_len(1284): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34658) uart: uart_read_bytes(1226): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34659) uart: uart_get_buffered_data_len(1284): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34659) uart: uart_read_bytes(1226): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34659) uart: uart_get_buffered_data_len(1284): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34660) uart: uart_read_bytes(1226): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34660) uart: uart_get_buffered_data_len(1284): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34660) uart: uart_read_bytes(1226): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34660) uart: uart_get_buffered_data_len(1284): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34661) uart: uart_read_bytes(1226): uart driver error

**Comment by adonno** [comment]
i'va had issues not sure if related or not. 

```
[00:52:56][D][esp-idf:000]: E (34658) uart: uart_get_buffered_data_len(1284): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34658) uart: uart_read_bytes(1226): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34659) uart: uart_get_buffered_data_len(1284): uart driver error
[00:52:56]
[00:52:56][D][esp-idf:000]: E (34659) uart: uart_read_bytes(1226): uart driver error
```

File: esphome/components/logger/__init__.py [hunk 1]
```diff
@@ -54,36 +59,51 @@
...
+UART0 = "UART0"
+UART1 = "UART1"
+UART2 = "UART2"
+UART0_SWAP = "UART0_SWAP"
+USB_SERIAL_JTAG = "USB_SERIAL_JTAG"
+USB_CDC = "USB_CDC"
+
+UART_SELECTION_ESP32 = {
+    VARIANT_ESP32: [UART0, UART1, UART2],
+    VARIANT_ESP32S2: [UART0, UART1, USB_SERIAL_JTAG],
```
Comment (jhamhader) [inline esphome/components/logger/__init__.py]: For the S2, it should be USB CDC.
For the C3, it should be USB_SERIAL_JTAG.
Not sure about S3.

Comment (jesserockz) [reply esphome/components/logger/__init__.py]: Haha, Oops. Thanks

**Pull request merged**